app/models/task.rb

```diff
@@ -39,29 +38,23 @@ def self.callback(msg)
     end
   end
 
-  def start!
-    self.options = task_options
-    if porter_enabled?
-      self.job_id = porter_start!(options)
-    end
-    save!
-  end
-
-  def task_options
-    {callback: callback_queue}.with_indifferent_access
+  def job_id
+    self[:job_id] ||= SecureRandom.uuid
   end
 
-  def feeder_storage_bucket
-    ENV["FEEDER_STORAGE_BUCKET"]
+  def source_url
   end
 
-  def callback_queue
-    q = "#{Rails.configuration.active_job.queue_name_prefix}_feeder_fixer_callback"
-    r = ENV["AWS_REGION"].present? ? ENV["AWS_REGION"] : "us-east-1"
-    "sqs://#{r}/#{q}"
-  end
-
-  def porter_enabled?
-    ENV["PORTER_SNS_TOPIC"].present?
+  def start!
+    self.status = "started"
+    self.options = {
+      Id: job_id,
+      Source: porter_source,
+      Tasks: porter_tasks,
+      Callbacks: porter_callbacks
+    }
+
+    porter_start!(options)
```

Comment:
i'm having trouble understanding this workflow. `porter_start!` starts a job, but i thought a job can contain tasks. so i'm a little confused on how this is being called from within a Task model/instance without causing an infinite call situation. unless jobs are started from Tasks?

Reply:
Terminology getting a bit overridden here... get ready for a long answer!

So in Feeder, we have a `Task` model in the DB.  Whose main purpose is to track a single [Porter Job](https://github.com/PRX/Porter#input-message-format).  And that Porter Job contains multiple Porter-Tasks.

So in the `CopyMediaTask`, we pull together the porter `job_id`, the `porter_source`, `porter_tasks` and `porter_callbacks` ... and assemble a chunk of JSON (saved to the db as `task.options`) which looks like this:

```json
{
  "Job": {
    "Id": "4ff6717f-a1e3-4f0f-a1a1-467a55026cce",
    "Source": {
      "Mode": "AWS/S3",
      "BucketName": "staging.mediajoint.prx.org",
      "ObjectKey": "public/audio_files/1212425/t121.mp3"
    },
    "Tasks": [
      {
        "Type": "Copy",
        "Mode": "AWS/S3",
        "BucketName": "staging-prx-feed",
        "ObjectKey": "878/4813ab85-902f-4eeb-967e-c39d7dc64c47/b5735ada-3662-4ea9-ad45-81ca5e53ee58.mp3",
        "ContentType": "REPLACE",
        "Parameters": {
          "CacheControl": "max-age=86400",
          "ContentDisposition": "attachment; filename=\"t121.mp3\""
        }
      },
      {
        "Type": "Inspect"
      }
    ],
    "Callbacks": [
      {
        "Type": "AWS/SQS",
        "Queue": "https://sqs.us-east-1.amazonaws.com/561178107736/4e489d22_stag_feeder_fixer_callback"
      }
    ]
  }
}
```

We also send that over [SNS](https://aws.amazon.com/sns/) (an AWS messaging thing ... like sending a text message) which triggers a [Porter execution](https://us-east-1.console.aws.amazon.com/states/home?region=us-east-1#/v2/executions/details/arn:aws:states:us-east-1:561178107736:execution:StateMachine-8B8z7vHLT4JS:a318d3b3-40cb-4f6c-9f6f-99d689ff015b).  (Porter is a "state machine" but that's maybe TMI for right now).  If you click on the "Execution input and output" tab you'll see that JSON.

So Porter performs those tasks (inspecting the source file, and copying it over to that location).  And then sends some callback JSON (you can also see it on the input-and-output tab) to the [AWS SQS](https://aws.amazon.com/sqs/) queue we indicated.

```json
{
  "Time": "2023-05-18T16:13:31.328Z",
  "Timestamp": 1684426411.328,
  "JobResult": {
    "Job": {
      "Id": "4ff6717f-a1e3-4f0f-a1a1-467a55026cce"
    },
    "Execution": {
      "Id": "arn:aws:states:us-east-1:561178107736:execution:StateMachine-8B8z7vHLT4JS:a318d3b3-40cb-4f6c-9f6f-99d689ff015b"
    },
    "FailedTasks": [],
    "State": "DONE",
    "TaskResults": [
      {
        "Task": "Copy",
        "Mode": "AWS/S3",
        "BucketName": "staging-prx-feed",
        "ObjectKey": "878/4813ab85-902f-4eeb-967e-c39d7dc64c47/b5735ada-3662-4ea9-ad45-81ca5e53ee58.mp3",
        "Time": "2023-05-18T16:13:30.206Z",
        "Timestamp": 1684426410.206
      },
      {
        "Task": "Inspect",
        "Inspection": {
          "Size": 482353,
          "Extension": "mp3",
          "MIME": "audio/mpeg",
          "Audio": {
            "Duration": 121051,
            "Format": "mp3",
            "Bitrate": 31878,
            "Frequency": 44100,
            "Channels": 2,
            "Layout": "stereo",
            "Layer": "3",
            "Frames": 4635
          }
        }
      }
    ]
  }
}
```

The Feeder worker is watching that SQS queue, sees the new message, picks it up, and calls [Task.callback](https://github.com/PRX/feeder.prx.org/blob/master/app/workers/porter_callback_worker.rb#L5) with the JSON.

That lands in the class method [self.callback](https://github.com/PRX/feeder.prx.org/blob/master/app/models/task.rb#L27) ... which tries to find a `Task` in the feeder db, with that Porter "Job Id" `4ff6717f-a1e3-4f0f-a1a1-467a55026cce`.

If found (and some other timestamp-check guards) ... we `task.update!(result: msg)](status: status, logged_at: time, result: msg)`.  Which drops us into the [before_save callback](https://github.com/PRX/feeder.prx.org/blob/master/app/models/tasks/copy_media_task.rb#L2)... which then parses that JSON result, pulling out the audio/image metadata.

Reply:
a little over my head but i think i kind of get the gist. a Task model in Feeder kind of represents a Porter Job, which itself consists of porter tasks, which are the individual tasks given to the Job by whichever Task model/instance defines those tasks, like `CopyImageTask` or `CopyMediaTask`.

Reply:
+1 you got it.  And a lot more succinct than my description!